### PR TITLE
fixed the hasApproval implementation

### PR DESCRIPTION
**Pull request by grodrigues3**
Now we allow for approval for any approver in path
Also allow recognize approve cancel comment

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1923)

### CODE REVIEW

**Comment by grodrigues3** [comment]
@foxish @apelisse @hongchaodeng 
Note that I tested this implementation on #1914 and it correct *adds and *removes the label as expect. *I ran this in dry-run mode

File: mungegithub/issue-labeler/OWNERS [hunk 1]
```diff
@@ -1,4 +1,4 @@
-assignees
+assignees:
```
Comment (hongchaodeng) [inline mungegithub/issue-labeler/OWNERS]: Can we put this into another commit?

File: mungegithub/mungers/approval-handler.go [hunk 2]
```diff
@@ -108,37 +106,28 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
 		}
 	}
+	glog.Infof("This is the cancel set %v", cancelSet)
 
+	//Checks that no valid approver has commented to cancel the approve label since last commit
+	//Checks that all files have an approver in the approverSet
 	for _, file := range files {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
+		fileOwners := h.features.Repos.Assignees(*file.Filename)
+		if fileOwners.Intersection(cancelSet).Len() > 0 {
```
Comment (hongchaodeng) [inline mungegithub/mungers/approval-handler.go]: Can you add the algorithm of cancelling into function comments too?
I think the comments doesn't reflect well enough what is being changed.

**Review by hongchaodeng — COMMENTED** [review]
Reviewed the approver handler.
Not sure about the "repo-updates.go".

Thanks @grodrigues3 for doing the work.

File: mungegithub/issue-labeler/OWNERS (hunk 1 above)
Comment (grodrigues3) [inline mungegithub/issue-labeler/OWNERS]: you mean separate PR or just a separate commit? It was mostly for testing purposes

Comment (hongchaodeng) [inline mungegithub/issue-labeler/OWNERS]: Commit

> It was mostly for testing purposes

Understand. That's why we should have separate commit.

**Comment by grodrigues3** [comment]
any other thoughts before apply lgtm label?  Anyone have reasonable ideas for automatically testing this?

**Comment by grodrigues3** [comment]
@apelisse can you PTAL?

File: mungegithub/mungers/approval-handler.go [hunk 3]
```diff
@@ -64,15 +61,17 @@ func (*ApprovalHandler) AddFlags(cmd *cobra.Command, config *github.Config) {}
 
 // Munge is the workhorse the will actually make updates to the PR
 // The algorithm goes as:
-// - Initially, we set up approverSet
-//   - Go through all comments after latest commit. If any approver said "/approve", add him to approverSet.
-// - For each file, we see if any approver of this file is in approverSet.
+// - Initially, we build two sets, an approverSet and a cancelSet
+//   - Go through all comments after latest commit.
+//	- If anyone said "/approve", add them to approverSet.
+//	- If anyone said "/approve cancel", add them to cancelSet.
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: So what happens if I do (in three different comments):
```
/approve
/approve cancel
/approve
```
First I'm in the `approverSet`, then I'm in `cancelSet`, and the last `/approve` is a no-op?

Suggested improvement:
- Add to `approverSet` when you see `/approve`.
- Remove from `approverSet` when you see `/approve cancel`
- Remove label if already applied and shouldn't be anymore

File: mungegithub/mungers/approval-handler.go [hunk 4]
```diff
@@ -64,15 +61,17 @@ func (*ApprovalHandler) AddFlags(cmd *cobra.Command, config *github.Config) {}
...
-//     - It's known that each dir has a list of approvers. (This might not hold true. For usability, current situation is enough.)
-//     - Approver of a dir is also the approver of child dirs.
-//   - We look at top N (default 3) level dir approvers. For example, for file "/a/b/c/d/e", we might search for approver from
-//     "/", "/a/", "/a/b/"
-// - Iff all files has been approved, the bot will add "approved" label.
+//     - Someone listed as an "approver" in an OWNERS file in the files directory OR
+//     - in one of the file's parent directories
+//   - If a valid approver is in cancelSet, we remove the approve label (if there is one) and return immediately
+// - Iff all files have been approved, the bot will add the "approved" label.
+// - This implies that if any file doesn't have "approved" from a valid approver, the approval label will not be applied
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Not for this PR, but we definitely need to tell people what files are not approved, and who can approve

File: mungegithub/mungers/approval-handler.go [hunk 5]
```diff
@@ -103,42 +103,30 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
-			continue
-		}
-
-		if len(fields) == 2 && strings.ToLower(fields[0]) == "/approve" && strings.ToLower(fields[1]) == "cancel" {
+		} else if len(fields) == 2 && strings.ToLower(fields[0]) == "/approve" && strings.ToLower(fields[1]) == "cancel" {
 			approverSet.Delete(*c.User.Login)
+			cancelSet.Insert(*c.User.Login)
 		}
 	}
+	glog.Infof("This is the cancel set %v", cancelSet)
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Please remove debug code :-)

File: mungegithub/mungers/approval-handler.go [hunk 6]
```diff
@@ -90,6 +90,7 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 	}
 
 	approverSet := sets.String{}
+	approverSet.Insert(*obj.Issue.User.Name)
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: BCNTC (Better check null than crash, I just made that up). Applies specifically well when the data is being returned by a third party interface.

File: mungegithub/mungers/approval-handler.go [hunk 7]
```diff
@@ -90,6 +90,7 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
 
 	// from oldest to latest
 	for i := len(comments) - 1; i >= 0; i-- {
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Can you please maybe break it into function for readability: `CreateApproverSet() sets.String` here instead of the loop

File: mungegithub/mungers/approval-handler.go [hunk 8]
```diff
@@ -103,42 +104,32 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+		} else if len(fields) == 2 && strings.ToLower(fields[0]) == "/approve" && strings.ToLower(fields[1]) == "cancel" {
+			if approverSet.Has(*c.User.Login) {
+				approverSet.Delete(*c.User.Login)
+			}
 		}
 	}
 
+	//Checks that no valid approver has commented to cancel the approve label since last commit
+	//Checks that all files have an approver in the approverSet
 	for _, file := range files {
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: New function: `AreAllFilesApproved() bool`.

(You're mixing the approval checking logic with the action that it implies)

File: mungegithub/mungers/approval-handler.go [hunk 9]
```diff
@@ -103,42 +104,32 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+			if obj.HasLabel(approvedLabel) {
+				glog.Infof("Canceling the approve label because %v file has no valid approvers", *file.Filename)
+				obj.RemoveLabel(approvedLabel)
+			}
 			return
 		}
 	}
+
+	//every file has been approved by a valid approver
 	obj.AddLabel(approvedLabel)
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Doesn't have to be in a function: if approved addLabel, else RemoveLabel.

Note that my patch needs to go in first if you don't want to check for the presence of the label before you add/remove.

File: mungegithub/mungers/approval-handler.go [hunk 10]
```diff
@@ -103,42 +104,39 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	needsApproval := sets.String{}
 	for _, file := range files {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
+		if !h.isApproved(file, approverSet) {
+			//no valid approvers
+			needsApproval.Insert(*file.Filename)
+			if obj.HasLabel(approvedLabel) {
+				obj.RemoveLabel(approvedLabel)
+			}
 			return
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Early return is wrong now

File: mungegithub/mungers/approval-handler.go [hunk 11]
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 	}
 
 	approverSet := sets.String{}
+	if obj.Issue.User.Name != nil {
+		approverSet.Insert(*obj.Issue.User.Name)
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: @fejta Automatically adding author to the `approverSet` should be the simplest way to make the process simple when you modify your own code/area.

**Review by apelisse — APPROVED** [review]
Awesome! To be honest, even though you made the logic more complicated than before, it's ten times easier to understand what it does ...

File: mungegithub/mungers/approval-handler.go [hunk 12]
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	}
 
 	// from oldest to latest
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Dangling comment

File: mungegithub/mungers/approval-handler.go [hunk 13]
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	approverSet = approverSet.Union(createApproverSet(comments))
+	//Checks that no valid approver has commented to cancel the approve label since last commit
+	//Checks that all files have an approver in the approverSet
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Dangling comments

File: mungegithub/mungers/approval-handler.go [hunk 14]
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	needsApproval := h.getApprovalNeededFiles(files, approverSet)
+
+	//every file has been approved by a valid approver
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Comment says the opposite (maybe just remove it)

File: mungegithub/mungers/approval-handler.go [hunk 15]
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	if needsApproval.Len() > 0 {
+		//need to selectively write a comment explaining why approved not added
+		glog.Infof("Canceling the approve label because these files %v have no valid approvers", needsApproval)
+		if obj.HasLabel(approvedLabel) {
+			obj.RemoveLabel(approvedLabel)
+		}
+	} else {
+		obj.AddLabel(approvedLabel)
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Above you test if the label is already there, here you don't check if it's already there or not. You probably should

File: mungegithub/mungers/approval-handler.go [hunk 16]
```diff
@@ -64,15 +62,17 @@ func (*ApprovalHandler) AddFlags(cmd *cobra.Command, config *github.Config) {}
 
 // Munge is the workhorse the will actually make updates to the PR
 // The algorithm goes as:
-// - Initially, we set up approverSet
-//   - Go through all comments after latest commit. If any approver said "/approve", add him to approverSet.
-// - For each file, we see if any approver of this file is in approverSet.
+// - Initially, we build two sets, an approverSet and a cancelSet
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Comments needs an update

File: mungegithub/mungers/approval-handler.go [hunk 17]
```diff
@@ -21,14 +21,21 @@ import (
...
 
+	"bytes"
+	"fmt"
 	"github.com/golang/glog"
 	goGithub "github.com/google/go-github/github"
 	"github.com/spf13/cobra"
 )
 
+const (
+	approvalNotificationName = "ApprovalNotifier"
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Not super fan of the name, maybe `[APPROVED-FILES]` ? But let's not bikeshed too much...

File: mungegithub/mungers/approval-handler.go [hunk 18]
```diff
@@ -109,6 +118,60 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+		glog.Error("Could not list the files for PR%v", obj.Issue.Number)
+		return
+	}
+	approvedFiles := sets.String{}
+	for _, fn := range files {
+		if !filesNeedApproval.Has(*fn.Filename) {
+			approvedFiles.Insert(*fn.Filename)
+		}
+	}
+	context := bytes.Buffer{}
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: https://golang.org/pkg/bytes/#NewBufferString

File: mungegithub/mungers/approval-handler.go [hunk 19]
```diff
@@ -109,6 +118,60 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	context.WriteString("The following files have been approved:\n")
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Not super fan of splitting the two lists

File: mungegithub/mungers/approval-handler.go [hunk 20]
```diff
@@ -109,6 +118,60 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	fileFmt := "\t%s"
+	for _, fn := range approvedFiles {
+		context.WriteString(fmt.Sprintf(fileFmt, fn))
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: should be the `OWNERS` file, not the file per se. Or we will have too many of these. And it'd be nice if it's a link. and if we can see who approved what ;-)

File: mungegithub/mungers/approval-handler.go [hunk 21]
```diff
@@ -90,8 +96,88 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
 	approverSet := sets.String{}
+	if obj.Issue.User.Name != nil {
+		approverSet.Insert(*obj.Issue.User.Name)
+	}
+
+	approverSet = approverSet.Union(createApproverSet(comments))
+	needsApproval := h.getApprovalNeededFiles(files, approverSet)
+
+	if needsApproval.Len() > 0 {
+		glog.Infof("Canceling the approve label because these files %v have no valid approvers", needsApproval)
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: This log doesn't seem right. It's not cancelling anything.

File: mungegithub/mungers/approval-handler.go [hunk 22]
```diff
@@ -90,8 +96,88 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+		//if latest notification stale, update it
+		if notificationNeedsUpdate(obj) {
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: By reading this code, I don't know if it removes it or not. Maybe rename the function?

File: mungegithub/mungers/approval-handler.go [hunk 23]
```diff
@@ -90,8 +96,88 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	if latestApprove.CreatedAt.Before(*lastModified) {
+		obj.DeleteComment(latestNotification)
+		return true
+	}
+	return false
+}
 
-	// from oldest to latest
+func createMessage(obj *github.MungeObject, filesNeedApproval sets.String) {
+	files, err := obj.ListFiles()
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: You could just pass-in files directly, you have it in the parent function

File: mungegithub/mungers/approval-handler.go (hunk 19 above)
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: But whatever :-)

File: mungegithub/mungers/approval-handler.go [hunk 24]
```diff
@@ -90,8 +96,88 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+		if notificationNeedsUpdate(obj) {
+			createMessage(obj, needsApproval)
+		}
+		if obj.HasLabel(approvedLabel) {
+			obj.RemoveLabel(approvedLabel)
+		}
+	} else if !obj.HasLabel(approvedLabel) {
+		obj.AddLabel(approvedLabel)
+	}
+
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: extra-space :-)

File: mungegithub/mungers/approval-handler.go (hunk 21 above)
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: 3 lines below, it removes the "approved" label.  Isn't that a cancel?

File: mungegithub/mungers/approval-handler.go (hunk 23 above)
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: sure, but I need the obj for Posting the notification below regardless

File: mungegithub/mungers/approval-handler.go (hunk 20 above)
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: you never explained how i get the OWNERs file

File: mungegithub/mungers/approval-handler.go (hunk 22 above)
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: can you suggest a newName? deleteOldNotificationsAndCheckForUpdate seems a little long?

File: mungegithub/mungers/approval-handler.go (hunk 21 above)
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: =))

File: mungegithub/mungers/approval-handler.go (hunk 22 above)
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: It probably does too much then. `findObsoleteNotification() *IssueComment`. And remove it outside the function.

File: mungegithub/mungers/approval-handler.go (hunk 23 above)
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: the name is wrong anyway. Can you return the message and post it from the calling function?

File: mungegithub/mungers/approval-handler.go (hunk 20 above)
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: I don't know either. One needs to think about it

File: mungegithub/mungers/approval-handler.go [hunk 25]
```diff
@@ -90,8 +97,117 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+		}
+		if obj.HasLabel(approvedLabel) {
+			obj.RemoveLabel(approvedLabel)
+		}
+	} else if !obj.HasLabel(approvedLabel) {
+		obj.AddLabel(approvedLabel)
+	}
+}
+
+func findObscoleteNotification(obj *github.MungeObject) *goGithub.IssueComment {
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: minor nit: obsolete is spelled obsolete

File: mungegithub/mungers/approval-handler.go [hunk 26]
```diff
@@ -90,8 +97,117 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	}
+	lastModified := obj.LastModifiedTime()
+	if latestApprove.CreatedAt.Before(*lastModified) {
+		// a commit has changed since the last approval
+		return latestNotification
+	}
+	return nil
+
+}
+func notificationNeedsUpdate(obj *github.MungeObject) bool {
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: duplicate?

File: mungegithub/mungers/approval-handler.go [hunk 27]
```diff
@@ -91,8 +97,90 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	approvedList := approvedFiles.List()
+	sort.Strings(approvedList)
+	needsApprovalList := filesNeedApproval.List()
+	sort.Strings(needsApprovalList)
+
+	context := bytes.NewBufferString("The following files have been approved:\n")
+	fileFmt := "- %s\n"
+	for _, fn := range approvedList {
+		context.WriteString(fmt.Sprintf(fileFmt, fn))
+	}
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Append extra `\n` after this list

File: mungegithub/mungers/approval-handler.go [hunk 28]
```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
 	for _, file := range files {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
-			return
+		// find owners files for files that have not yet been approved
+		if !filesNeedApproval.Has(*file.Filename) {
+			ownersFiles.Insert(h.features.Repos.FindOwnersForPath(*file.Filename))
 		}
 	}
-	obj.AddLabel(approvedLabel)
+	context := bytes.NewBufferString("The PR requires approval from at least one person in the following OWNERS files:\n")
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Oh so you removed the list of already approved files?

File: mungegithub/mungers/approval-handler.go [hunk 29]
```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+		glog.Error("Could not list the files for PR%v", obj.Issue.Number)
+		return err
+	}
+	// contains the a set of directories where relevant ownersFiles are
+	ownersFiles := sets.String{}
 	for _, file := range files {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
-			return
+		// find owners files for files that have not yet been approved
+		if !filesNeedApproval.Has(*file.Filename) {
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: You should do this directly in `getApprovalNeededFiles`

Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: So if a file is NOT in `filesNeedApproval`, we looks for it `OWNERS` path? that seems wrong.

File: mungegithub/mungers/approval-handler.go [hunk 30]
```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	for _, file := range files {
+		if !h.isApproved(file, approverSet) {
+			needsApproval.Insert(*file.Filename)
+		}
+	}
+	return needsApproval
+}
+
+// isApproved indicates whether or not someone from the list of OWNERS for a file has approved the PR
+func (h ApprovalHandler) isApproved(someFile *goGithub.CommitFile, approverSet sets.String) bool {
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: This could operate directly on the list of OWNERS file instead, but whatever

File: mungegithub/mungers/approval-handler.go [hunk 31]
```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
-	return false
+	return approverSet
+}
+
+// getApprovalNeededFiles identifies the files that still need approval from someone in their OWNERS files
+func (h ApprovalHandler) getApprovalNeededFiles(files []*goGithub.CommitFile, approverSet sets.String) sets.String {
+	needsApproval := sets.String{}
+	for _, file := range files {
+		if !h.isApproved(file, approverSet) {
+			needsApproval.Insert(*file.Filename)
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: This should probably generate the OWNERS file path directly

File: mungegithub/mungers/approval-handler.go [hunk 32]
```diff
@@ -17,20 +17,24 @@ limitations under the License.
...
 	"k8s.io/contrib/mungegithub/github"
+	mungeComment "k8s.io/contrib/mungegithub/mungers/matchers/comment"
 	"k8s.io/contrib/mungegithub/mungers/mungerutil"
 	"k8s.io/kubernetes/pkg/util/sets"
 
+	"bytes"
+	"fmt"
 	"github.com/golang/glog"
-	githubapi "github.com/google/go-github/github"
+	goGithub "github.com/google/go-github/github"
```
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: deleting your comment

File: mungegithub/mungers/approval-handler.go (hunk 28 above)
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: yeah, just shows the OWNERS files that need an approver to sign off

File: mungegithub/features/repo-updates.go [hunk 33]
```diff
@@ -184,15 +172,15 @@ func (o *RepoInfo) updateRepoUsers() error {
 	}
 	sha := out
 
-	o.assignees = map[string]sets.String{}
-	//o.owners = map[string]sets.String{}
+	o.approvers = map[string]sets.String{}
```
Comment (grodrigues3) [inline mungegithub/features/repo-updates.go]: should be a make here?  And/or set.NewString{}

File: mungegithub/mungers/approval-handler.go [hunk 34]
```diff
@@ -90,62 +95,157 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
-	for i := 0; i < len(paths) && i < depth; i++ {
-		fileOwners := h.features.Repos.LeafAssignees(p)
-		if fileOwners.Len() == 0 {
-			glog.Warningf("Couldn't find an owner for path (%s)", p)
+// createApproverSet iterates through the list of comments on a PR
+// and identifies all of the people that have said /approve and adds
+// them to the approverSet.  The function uses the latest approve or cancel comment
+// to determine the Users intention
+func createApproverSet(comments []*goGithub.IssueComment) sets.String {
+	approverSet := sets.String{}
```
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: sets.NewString()

File: mungegithub/mungers/approval-handler.go [hunk 35]
```diff
@@ -90,62 +95,157 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
-		p = path.Join(p, paths[i])
 	}
-	return false
+	return approverSet
+}
+
+// getApprovalNeededFiles identifies the files that still need approval from someone in their OWNERS files
+func (h ApprovalHandler) getApprovalNeededOwnersFiles(files []*goGithub.CommitFile, approverSet sets.String) (sets.String, sets.String) {
+	owners := sets.String{}
+	hasApproval := sets.String{}
```
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: sets.NewString()

File: mungegithub/mungers/approval-handler.go [hunk 36]
```diff
@@ -90,62 +95,157 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+				approverCount[approver] = sets.NewString(ownersFile)
+			}
 		}
 	}
 
-	for _, file := range files {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
-			return
+	var copyOfFiles sets.String
+	for fn := range ownersPath {
```
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: uninitialized variable

File: mungegithub/features/repo-updates.go (hunk 33 above)
Comment (apelisse) [inline mungegithub/features/repo-updates.go]: Nah, I think this is fine

File: mungegithub/mungers/approval-handler.go [hunk 37]
```diff
@@ -17,20 +17,25 @@ limitations under the License.
...
 	"k8s.io/contrib/mungegithub/github"
+	mungeComment "k8s.io/contrib/mungegithub/mungers/matchers/comment"
 	"k8s.io/contrib/mungegithub/mungers/mungerutil"
 	"k8s.io/kubernetes/pkg/util/sets"
 
+	"bytes"
+	"fmt"
 	"github.com/golang/glog"
-	githubapi "github.com/google/go-github/github"
+	goGithub "github.com/google/go-github/github"
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: as agreed ;-)

File: mungegithub/mungers/approval-handler.go [hunk 38]
```diff
@@ -90,62 +96,165 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
+	for fn := range ownersPath {
+		copyOfFiles.Insert(fn)
+	}
+
+	approverGroup := sets.NewString()
+	for copyOfFiles.Len() > 0 {
+		maxCovered := 0
+		var bestPerson string
+		for k, v := range approverCount {
+			if len(v) >= maxCovered && v.Intersection(copyOfFiles).Len() != 0 {
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: why isn't this:
```
v.Intersection(copyOfFiles).Len() > maxCovered
```

Why `>=` and why `len(v)` rather than intersection?

File: mungegithub/mungers/approval-handler.go [hunk 39]
```diff
@@ -90,62 +95,169 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
-		if !mungerutil.IsValidUser(c.User) {
-			continue
+func (h *ApprovalHandler) updateNotification(obj *github.MungeObject, approvedFiles, needsApprovalOwnersFiles sets.String) error {
+	notificationMatcher := mungeComment.MungerNotificationName(approvalNotificationName)
+	comments, err := obj.ListComments()
+	if err != nil {
+		glog.Error("Could not list the comments for PR%v", obj.Issue.Number)
+		return err
+	}
+	latestNotification := mungeComment.FilterComments(comments, notificationMatcher).GetLast()
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Doesn't look like me, somebody must have forced me to do that during my sleep :-)

File: mungegithub/mungers/approval-handler.go [hunk 40]
```diff
@@ -90,62 +96,165 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
...
-	for _, file := range files {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
-			return
+func (h *ApprovalHandler) createMessage(obj *github.MungeObject, ownersMap map[string]sets.String) error {
+	//sort the keys so we always display OWNERS files in same order
+	sliceOfKeys := make([]string, len(ownersMap))
+	i := 0
+	for path := range ownersMap {
+		sliceOfKeys[i] = path
+	}
```
Comment (grodrigues3) [inline mungegithub/mungers/approval-handler.go]: keeping my implementation (with incr fix)

**Comment by apelisse** [comment]
Looks good to me! :-)
/lgtm

I'll take care of squashing.

**Comment by grodrigues3** [comment]
#2076
